### PR TITLE
cpu/stm32: add ADCAL operation

### PR DESCRIPTION
**Pull request by david-vankampen**
The STM32G0 has an ADC Calibration procedure ad described in RM0454 Section 14.3.3
image

### Contribution description

Add the software calibration procedure on startup as described in STM's documentation.

### CODE REVIEW

File: cpu/stm32/periph/adc_f0_g0.c [hunk 1]
```diff
@@ -75,6 +75,19 @@ int adc_init(adc_t line)
     }
     /* reset configuration */
     ADC1->CFGR2 = 0;
+    ADC1->CR |= ADC_CR_ADVREGEN;
+    ADC1->CR &= ~ADC_CR_ADCAL;
+    // calibrate ADC, per RM0454 section 14.3.3
+    // 1. ensure ADEN=0, ADVREGEN=1, DMAEN=0
+    if (!(ADC1->CR & ADC_CR_ADEN) && (ADC1->CR & ADC_CR_ADVREGEN) && !(ADC1->CFGR1 & ADC_CFGR1_DMAEN))
```
Comment (benpicco) [inline cpu/stm32/periph/adc_f0_g0.c:82]: Under which circumstances would those not be set?
You are already setting `ADVREGEN` in the line above, so why not also

```C
ADC1->CR &= ~(ADC_CR_ADCAL | ADC_CR_ADEN | ADC_CFGR1_DMAEN);
```

File: cpu/stm32/periph/adc_f0_g0.c [hunk 2]
```diff
@@ -75,6 +75,19 @@ int adc_init(adc_t line)
...
+    {
+        // 2. Set ADCAL=1
+        ADC1->CR |= ADC_CR_ADCAL;
+        // 3. Wait for ADCAL=0 (or EOCAL=1)
+        while ((ADC1->ISR & ADC_ISR_EOCAL))
+        {
+        }
```
Comment (benpicco) [inline cpu/stm32/periph/adc_f0_g0.c:89]:
```suggestion
        while ((ADC1->ISR & ADC_ISR_EOCAL)) {}
```

**Review by benpicco — COMMENTED** [review]
Just two style nits:

 - also prefix the commit message with `cpu/stm32` like you did for the PR title
 - use `/* this comment style */` instead of `// this comment style` for consistency.

**Comment by aabadie** [comment]
This doesn't build on stm32f0 based boards and the ADC implementation is shared between f0 and g0:

```
make BOARD=nucleo-f072rb -C tests/periph/adc/
make: Entering directory '/work/riot/RIOT/tests/periph/adc'
Building application "tests_adc" for "nucleo-f072rb" with MCU "stm32".

"make" -C /work/riot/RIOT/pkg/cmsis/ 
"make" -C /work/riot/RIOT/boards/common/init
"make" -C /work/riot/RIOT/boards/nucleo-f072rb
"make" -C /work/riot/RIOT/boards/common/nucleo
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/core/lib
"make" -C /work/riot/RIOT/cpu/stm32
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32/periph
/work/riot/RIOT/cpu/stm32/periph/adc_f0_g0.c: In function 'adc_init':
/work/riot/RIOT/cpu/stm32/periph/adc_f0_g0.c:78:17: error: 'ADC_CR_ADVREGEN' undeclared (first use in this function); did you mean 'ADC_CCR_VREFEN'?
   78 |     ADC1->CR |= ADC_CR_ADVREGEN;
      |                 ^~~~~~~~~~~~~~~
      |                 ADC_CCR_VREFEN
/work/riot/RIOT/cpu/stm32/periph/adc_f0_g0.c:78:17: note: each undeclared identifier is reported only once for each function it appears in
/work/riot/RIOT/cpu/stm32/periph/adc_f0_g0.c:88:29: error: 'ADC_ISR_EOCAL' undeclared (first use in this function); did you mean 'ADC_ISR_EOC'?
   88 |         while ((ADC1->ISR & ADC_ISR_EOCAL)) {}
      |                             ^~~~~~~~~~~~~
      |                             ADC_ISR_EOC
make[3]: *** [/work/riot/RIOT/Makefile.base:146: /work/riot/RIOT/tests/periph/adc/bin/nucleo-f072rb/periph/adc_f0_g0.o] Error 1
make[2]: *** [/work/riot/RIOT/Makefile.base:31: ALL--/work/riot/RIOT/cpu/stm32/periph] Error 2
make[1]: *** [/work/riot/RIOT/Makefile.base:31: ALL--/work/riot/RIOT/cpu/stm32] Error 2
make: *** [/work/riot/RIOT/tests/periph/adc/../../../Makefile.include:761: application_tests_adc.module] Error 2
make: Leaving directory '/work/riot/RIOT/tests/periph/adc'
```

I haven't checked the f0 datasheet to see if the changes in the PR could apply to f0 (which I doubt) so maybe just add some
`#if defined(ADC_CR_ADVREGEN)/#endif` around the code added would be enough.

File: cpu/stm32/periph/adc_f0_g0.c (hunk 1 above)
Comment (benpicco) [reply cpu/stm32/periph/adc_f0_g0.c:84]: I think with 'ensure' they mean you should set those bits

File: cpu/stm32/periph/adc_f0_g0.c (hunk 2 above)
Comment (david-vankampen) [reply cpu/stm32/periph/adc_f0_g0.c]: I haven't timed it, but it exited immediately in the brief testing I did.

**Review by benpicco — APPROVED** [review]
I didn't test this yet, but code looks good.
Can you provide some results?

**Comment by david-vankampen** [comment]
@benpicco I don't have hard numbers with me, but this change, coupled with the adjustment using Vrefint (to compensate for a moving or unknown Vref) got us an improvement of about 50-100mV in accuracy on the ADC reading. I don't have details to attribute which change provided more improvements, however.

**Comment by maribu** [comment]
I think this needs a rebase due to the migration from bors to Github merge queues. @david-vankampen would you mind to rebase this on current master?

**Comment by david-vankampen** [comment]
@maribu looks like this should now be all set on that front?

**Comment by maribu** [comment]
> @[github-merge-queue](https://github.com/apps/github-merge-queue) github-merge-queue bot removed this pull request from the [merge queue](https://github.com/RIOT-OS/RIOT/queue/master) due to no response for status checks Nov 28, 2023

Fantastic, just minutes before the build was completed :cry:

**Pull request merged**